src/network/network_crypto_internal.h

```diff
@@ -0,0 +1,317 @@
+
+/** Container for a X25519 key exchange message. */
+using X25519KeyExchangeMessage = std::array<uint8_t, X25519_KEY_EXCHANGE_MESSAGE_SIZE>;
+
+/** The side of the key exchange. */
+enum class X25519KeyExchangeSide {
+	CLIENT, ///< We are the client.
+	SERVER, ///< We are the server.
+};
+
+/**
+ * Container for the keys that derived from the X25519 key exchange mechanism. This mechanism derives
+ * a key to encrypt both the client-to-server and a key to encrypt server-to-client communication.
+ */
+class X25519DerivedKeys {
+private:
+	/** Single contiguous buffer to store the derived keys in, as they are generated as a single hash. */
+	std::array<uint8_t, X25519_KEY_SIZE + X25519_KEY_SIZE> keys;
+public:
+	~X25519DerivedKeys();
+	std::span<const uint8_t> ClientToServer() const;
+	std::span<const uint8_t> ServerToClient() const;
+	bool Exchange(const X25519PublicKey &peer_public_key, X25519KeyExchangeSide side,
+			const X25519SecretKey &our_secret_key, const X25519PublicKey &our_public_key, std::string_view extra_payload);
+};
+
+/**
+ * Base for handlers using a X25519 key exchange to perform authentication.
+ *
+ * In general this works as follows:
+ * 1) the client and server have or generate a secret and public X25519 key.
+ * 2) the X25519 key exchange is performed at both the client and server, with their own secret key and their peer's public key.
+ * 3) a pair of derived keys is created by BLAKE2b-hashing the following into 64 bytes, in this particular order:
+ *    - the shared secret from the key exchange;
+ *    - the public key of the server;
+ *    - the public key of the client;
+ *    - optional extra payload, e.g. a password in the case of PAKE.
+ *    The first of the pair of derived keys is usually used to encrypt client-to-server communcation, and the second of the pair
+ *    is usually used to encrypt server-to-client communication.
+ * 4) a XChaCha20-Poly1305 (authenticated) encryption is performed using:
+ *    - the first of the pair of derived keys as encryption key;
+ *    - a 24 byte nonce;
+ *    - the public key of the client as additional authenticated data.
+ *    - a 8 byte random number as content/message.
+ *
+ * The server initiates the request by sending its public key and a 24 byte nonce that is randomly generated. Normally the side
+ * that sends the encrypted data sends the nonce in their packet, which would the the client on our case. However, there are
+ * many implementations of clients due to the admin-protocol where this is used, and we cannot guarantee that they generate a
+ * good enough nonce. As such the server sends one instead. The server will create a new set of keys for each session.
+ *
+ * The client receives the request, performs the key exchange, generates the derived keys and then encrypts the message. This
+ * message must contain some content, so it has to be filled with a four random bytes. Once the message has been encrypted, the
```

Comment:
```suggestion
 * message must contain some content, so it has to be filled with a 8 random bytes. Once the message has been encrypted, the
```

At least, you say 8 a bit higher.